runtime/v2/runc/task/service.go

```diff
@@ -517,48 +523,63 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-func (s *service) sendL(evt interface{}) {
-	s.eventSendMu.Lock()
-	s.events <- evt
-	s.eventSendMu.Unlock()
-}
-
 func (s *service) checkProcesses(e runcC.Exit) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	missing := make(map[process.Process]*runc.Container)
 	for _, container := range s.containers {
-		if !container.HasPid(e.Pid) {
-			continue
-		}
-
 		for _, p := range container.All() {
+			if _, starting := s.processEventsLocker.Load(p.ID()); starting {
+				missing[p] = container
+				continue
+			}
 			if p.Pid() != e.Pid {
 				continue
 			}
 
-			if ip, ok := p.(*process.Init); ok {
-				// Ensure all children are killed
-				if runc.ShouldKillAllOnExit(s.context, container.Bundle) {
-					if err := ip.KillAll(s.context); err != nil {
-						log.L.WithError(err).WithField("id", ip.ID()).
-							Error("failed to kill init's children")
-					}
-				}
+			s.sendProcessEvent(e, p, container)
+			return
+		}
+	}
+
+	tryMissing := func(e runcC.Exit, p process.Process, c *runc.Container) {
+		for {
+			if _, starting := s.processEventsLocker.Load(p.ID()); !starting {
+				break
 			}
+			time.Sleep(50 * time.Millisecond)
```

Comment:
We can do better than polling. Store a channel in `s.processEventsLocker` and close it when deleting the entry from the map. Then this function can receive on the channel to wait.
```go
if startedCh, ok := s.processEventsLocker.Load(p.ID()); ok {
	<-startedCh.(chan struct{})
}
```